File: packages/capacitor/android/src/main/kotlin/app/quiltt/capacitor/QuilttConnectorPlugin.kt
package app.quiltt.capacitor

import android.content.Intent
import android.net.Uri
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

/**
 * QuilttConnector Capacitor Plugin for Android
 * Handles deep linking, URL opening, and OAuth redirect flows for Quiltt Connector integration
 */
@CapacitorPlugin(name = "QuilttConnector")
class QuilttConnectorPlugin : Plugin() {
    private var appLauncherUrl: String? = null

    override fun load() {
        // Capture the app launcher URL when the app starts from a deep-link intent
        val intent = activity?.intent
        handleIntent(intent)
    }

    override fun handleOnNewIntent(intent: Intent) {
        super.handleOnNewIntent(intent)
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent?) {
        val data = intent?.data
        if (data != null) {
            val url = data.toString()
            appLauncherUrl = url

            // Notify JavaScript listeners
            val ret = JSObject()
            ret.put("url", url)
            notifyListeners("deepLink", ret)
        }
    }

    /**
     * Opens a URL in the system browser
     * Used for OAuth flows and external authentication
     */
    @PluginMethod
    fun openUrl(call: PluginCall) {
        val urlString = call.getString("url")
        if (urlString.isNullOrEmpty()) {
            call.reject("Invalid URL")
            return
        }

        try {
            val uri = Uri.parse(urlString)
            val intent = Intent(Intent.ACTION_VIEW, uri)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            context.startActivity(intent)

            val ret = JSObject()
            ret.put("completed", true)
            call.resolve(ret)
        } catch (e: Exception) {
            call.reject("Failed to open URL: ${e.message}")
        }
    }

    /**
     * Returns the app launcher URL (if any)
     * Used to handle OAuth callbacks and deep link navigation
     */
    @PluginMethod
    fun getAppLauncherUrl(call: PluginCall) {
        val ret = JSObject()
        if (appLauncherUrl != null) {
            ret.put("url", appLauncherUrl)
        } else {
            ret.put("url", JSObject.NULL)
        }
        call.resolve(ret)
    }
}

File: packages/capacitor/ios/Sources/QuilttConnectorPlugin/QuilttConnectorPlugin.swift
import Foundation
import Capacitor

/// QuilttConnector Capacitor Plugin for iOS
/// Handles deep linking, URL opening, and OAuth redirect flows for Quiltt Connector integration
@objc(QuilttConnectorPlugin)
public class QuilttConnectorPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "QuilttConnectorPlugin"
    public let jsName = "QuilttConnector"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "openUrl", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getLaunchUrl", returnType: CAPPluginReturnPromise)
    ]

    private var launchUrl: URL?

    public override func load() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleOpenUrl(_:)),
            name: Notification.Name.capacitorOpenURL,
            object: nil
        )
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleOpenUrl(_:)),
            name: Notification.Name.capacitorOpenUniversalLink,
            object: nil
        )
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    /// Handle incoming URLs from deep links or universal links
    @objc private func handleOpenUrl(_ notification: Notification) {
        guard let object = notification.object as? [String: Any],
              let url = object["url"] as? URL else {
            return
        }

        // Store as launch URL
        launchUrl = url

        // Notify JavaScript listeners
        notifyListeners("deepLink", data: [
            "url": url.absoluteString
        ])
    }

    /// Opens a URL in the system browser
    /// Used for OAuth flows and external authentication
    @objc func openUrl(_ call: CAPPluginCall) {
        guard let urlString = call.getString("url"),
              let url = URL(string: urlString) else {
            call.reject("Invalid URL")
            return
        }

        DispatchQueue.main.async {
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(url, options: [:]) { success in
                    if success {
                        call.resolve([
                            "completed": true
                        ])
                    } else {
                        call.reject("Failed to open URL")
                    }
                }
            } else {
                let success = UIApplication.shared.openURL(url)
                if success {
                    call.resolve([
                        "completed": true
                    ])
                } else {
                    call.reject("Failed to open URL")
                }
            }
        }
    }

    /// Returns the URL that launched the app (if any)
    /// Used to handle OAuth callbacks and deep link navigation
    @objc func getLaunchUrl(_ call: CAPPluginCall) {
        if let url = launchUrl {
            call.resolve([
                "url": url.absoluteString
            ])
        } else {
            call.resolve([
                "url": NSNull()
            ])
        }
    }
}
